metagpt/tools/libs/software_development.py

```diff
@@ -0,0 +1,268 @@
+    from metagpt.roles import Architect
+
+    ctx = Context()
+    ctx.set_repo_dir(project_path)
+
+    role = Architect(context=ctx)
+    await role.run(with_message=Message(content="", cause_by=WritePRD))
+    return project_path
+
+
+@register_tool(tags=["software company", "Architect"])
+async def write_project_plan(project_path: str | Path) -> Path:
+    """Writes a project plan to the project repository, based on the design of the project.
+
+    Args:
+        project_path (str|Path): The path to the project repository.
+
+    Returns:
+        Path: The path to the project directory
+
+    Example:
+        >>> from metagpt.tools.libs.software_development import write_project_plan
+        >>> project_path = '/path/to/project_path' # Returned by `write_prd`
+        >>> project_path = await write_project_plan(project_path)
+        >>> print(project_path)
+        '/path/to/project_path'
+
+    """
+    from metagpt.actions import WriteDesign
+    from metagpt.context import Context
+    from metagpt.roles import ProjectManager
+
+    ctx = Context()
+    ctx.set_repo_dir(project_path)
+
+    role = ProjectManager(context=ctx)
+    await role.run(with_message=Message(content="", cause_by=WriteDesign))
+    return project_path
+
+
+@register_tool(tags=["software company", "Engineer"])
+async def write_codes(project_path: str | Path, inc: bool = False) -> Path:
+    """Writes codes to the project repository, based on the project plan of the project.
+
+    Args:
+        project_path (str|Path): The path to the project repository.
+        inc (bool, optional): Whether to write incremental codes. Defaults to False.
+
+    Returns:
+        Path: The path to the project directory
+
+    Example:
+        # Write codes to a new project
+        >>> from metagpt.tools.libs.software_development import write_codes
+        >>> project_path = '/path/to/project_path' # Returned by `write_prd`
+        >>> project_path = await write_codes(project_path)
+        >>> print(project_path)
+        '/path/to/project_path'
+
+        # Write increment codes to the exists project
+        >>> from metagpt.tools.libs.software_development import write_codes
+        >>> project_path = '/path/to/project_path' # Returned by `write_prd`
+        >>> project_path = await write_codes(project_path, inc=True)
+        >>> print(project_path)
+        '/path/to/project_path'
+    """
+    from metagpt.actions import WriteTasks
+    from metagpt.context import Context
+    from metagpt.roles import Engineer
+
+    ctx = Context()
+    ctx.config.inc = inc
+    ctx.set_repo_dir(project_path)
+
+    role = Engineer(context=ctx)
+    msg = Message(content="", cause_by=WriteTasks, send_to=role)
+    me = {any_to_str(role), role.name}
+    while me.intersection(msg.send_to):
+        msg = await role.run(with_message=msg)
+    return project_path
+
+
+@register_tool(tags=["software company", "QaEngineer"])
+async def run_qa_test(project_path: str | Path) -> Path:
```

Comment:
A clear upstream file name is required, and a clear return file name is required. The current input and output granularity is too large

Reply:
fixed, input src path and return test path